### ISSUE TITLE
kernel profiles for osi_linux depend on qemu runtime configuration

### ISSUE THREAD

**Issue by m000** [panda-re/panda#476]
`osi_linux` plugin uses [kernel profiles](/panda-re/panda/blob/master/panda/plugins/osi_linux/kernelinfo.conf) in order to facilitate kernel state introspection. The profiles contain kernel-specific offsets of different kernel struct members. Profiles were meant to be extracted once for each kernel build and be shared and reused. However, this is no longer possible after the merge #461. 

#461 updates the code that retrieves the current `struct task_struct` address from the per-cpu storage area in the memory  via `__per_cpu_offset[0]`. This enables the (long awaited) support for 64bit linux variants. **However, it has the side-effect that kernel profiles are now dependent on the runtime configuration of the guest VM.** 

This is because per-cpu storage is dynamically allocated, thus affected by how the allocator behaves in different configurations. For example, the following 4 values of `__per_cpu_offset[0]` were acquired from the same VM image using 4 different hardware configurations:
```
# 512 MB - no kvm
#task.per_cpu_offset_0 = 0x1DE18000
task.per_cpu_offset_0 = 501317632

# 512 MB - kvm
#task.per_cpu_offset_0 = 0x1DE17000
task.per_cpu_offset_0 = 501313536

# 4096 MB - kvm
#task.per_cpu_offset_0 = 0x34129000
task.per_cpu_offset_0 = 873631744

# 2048 MB - no kvm
#task.per_cpu_offset_0 = 0x34B2A000
task.per_cpu_offset_0 = 884121600
```

This issue complicates the handling of PANDA recordings, as we now need to also associate a kernel profile with them. Previously, for pre-compiled kernels, knowing the version and the distro was sufficient.

**Comment by m000** [comment]
Ok, it seems that the problem is easier than it seemed on first glance: Instead of including directly 
`__per_cpu_offset[0]` in the profile, we should include `&__per_cpu_offset` instead. Then we dynamically read `__per_cpu_offset[0]` when we load the plugin. That should do it, right?

**Comment by nathanjackson** [comment]
That seems like the right solution to me. Although I think you probably want to read the offset on the first request to osi_linux and cache it, not in init_plugin. When init_plugin is called, I don't think there is any guarantee about the guest being "ready" for its memory to be read.

**Comment by m000** [comment]
I'm halfway implementing this solution. Should be able to push later tonight.

`__per_cpu_offset[0]` will still part of the kernel profile. This is to keep all the profile information at one location. But if `&__per_cpu_offset` is included in the profile (i.e. non-zero), then it will be updated to reflect the proper value.

**Comment by m000** [comment]
Closing. Resolved after merge of #474.
